Resolve dependencies in setup-ide step

tpasternak opened the pull request:
Fail only when dependencies passed via CLI cannot be resolved

Fail when someone passes '--dependency <invalid_dependency>
Do not fail when there is an invalid '$ivy.<dep>' in code

tpasternak committed:
Resolve dependencies in setup-ide step

Run dependencies resolution if any `--dependency` flags is passed to
the CLI

alexarchambault reviewed (commented):
Just pushed a commit with minor adjustments (using `options.buildOptions.classPathOptions.extraDependencies` rather than `options.shared.dependencies.dependency` - the latter might have empty strings, which are ignored, while the former doesn't).

Maybe some tests could be added. There's already a `setup-ide` test in [`BspTestDefinitions.scala`](https://github.com/VirtuslabRnD/scala-cli/blob/23571f78d9cbf81316ba57f8c3299204e3b8791e/modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala). We could ensure that a faulty dependency on the command-line makes `setup-ide` fails, while the same as an `import $ivy` doesn't.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala:
```diff
@@ -125,6 +125,38 @@ abstract class BspTestDefinitions(val scalaVersionOpt: Option[String])
+        "--dependency",
+        "com.lihaoyi::cask:0.7.11"
+      ).call(cwd = root, check = false)
+      expect(p.exitCode == 1)
```
Isn't it supposed to succeed here?
```suggestion
      os.proc(
        TestUtil.cli,
        "setup-ide",
        extraOptions,
        "--dependency",
        "com.lihaoyi::cask:0.7.11"
      ).call(cwd = root)
```

tpasternak replied:
Thanks for finding this, the tests were failing, because in both cases the setup-cli command was failing earlier ("No inputs found"). Now I think it's correct. I also had to adapt to `CrossSources` returning `Either` after rebase.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala:
```diff
@@ -149,6 +149,40 @@ abstract class BspTestDefinitions(val scalaVersionOpt: Option[String])
 
+  val importPprintOnlyProject = TestInputs(
+    Seq(
+      os.rel / "simple.sc" -> s"import $$ivy.`com.lihaoyi::pprint:${Constants.pprintVersion}`"
```
No need to get this one via constants IMO… It's fine to hardcode dependency versions in the tests.

tpasternak replied:
I did it to make it easy to track all the dependencies that are used during the build/testing. If we will use a tool to scan vulnerable dependencies in future, it won't find the hardcoded ones

tpasternak replied:
(but ofc I can revert it if you have a strong opinion)

alexarchambault replied:
There's already many hardcoded dependencies in the tests… 

Let's either revert, or put test dependencies aside from the main ones, in the build.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala:
```diff
@@ -149,6 +149,40 @@ abstract class BspTestDefinitions(val scalaVersionOpt: Option[String])
+        ".",
+        extraOptions,
+        "--dependency",
+        s"org.scalameta::munit:${Constants.munitVersion}"
```
Same as above, no need to get this version via the build.

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala:
```diff
@@ -149,6 +149,40 @@ abstract class BspTestDefinitions(val scalaVersionOpt: Option[String])
+        "--dependency",
+        s"org.scalameta::munit:${Constants.munitVersion}"
+      ).call(cwd = root, check = false)
+      expect(p.exitCode == 0)
```
Just removing `check = false` ought to be enough, no need to check for a `0` exit code.

tpasternak replied:
Done

tpasternak committed:
Throw exception instead of checking exitcode in tests

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala:
```diff
@@ -149,6 +149,39 @@ abstract class BspTestDefinitions(val scalaVersionOpt: Option[String])
+
+  test("setup-ide should succeed for valid dependencies") {
+    importPprintOnlyProject.fromRoot { root =>
+      val p = os.proc(
```
NIT
```suggestion
      os.proc(
```

alexarchambault commented on modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala:
```diff
@@ -149,6 +149,39 @@ abstract class BspTestDefinitions(val scalaVersionOpt: Option[String])
+        "--dependency",
+        "org.scalameta::munit:0.7.119999"
+      ).call(cwd = root, check = false)
+      expect(p.exitCode == 1)
```
By passing `stderr = os.Pipe` to `call`, we can get the errors printed by scala-cli, and ensure that the error originate from a not found dependency and not something else.

tpasternak committed:
Test: Verify whether setup-ide fails because of failed deps resolution

tpasternak committed:
Move some deps from Deps object to TestDeps

Just to separate dependencies that are used in tests execution from
the ones used to build scala-cli

alexarchambault reviewed (approved):
Let's merge if that's fine for you, @tpasternak.
